File: practice-projects/gin-mvc/internal/models/order/order.go
package order

import (
	"errors"
	"time"
)

type Status string

const (
	StatusPending   Status = "PENDING"
	StatusPaid      Status = "PAID"
	StatusShipped   Status = "SHIPPED"
	StatusDelivered Status = "DELIVERED"
	StatusCancelled Status = "CANCELLED"
	StatusRefunded  Status = "REFUNDED"
)

type Order struct {
	ID          int64     `json:"id"`
	OrderNo     string    `json:"order_no"`
	UserID      int64     `json:"user_id"`
	TotalAmount float64   `json:"total_amount"`
	Status      Status    `json:"status"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

func New(orderNo string, userID int64, totalAmount float64) (*Order, error) {
	if orderNo == "" {
		return nil, errors.New("order_no is required")
	}
	if userID <= 0 {
		return nil, errors.New("user_id is invalid")
	}
	if totalAmount <= 0 {
		return nil, errors.New("total_amount is invalid")
	}
	now := time.Now()
	return &Order{
		OrderNo:     orderNo,
		UserID:      userID,
		TotalAmount: totalAmount,
		Status:      StatusPending,
		CreatedAt:   now,
		UpdatedAt:   now,
	}, nil
}

func (o *Order) Pay() error {
	if o.Status != StatusPending {
		return errors.New("only pending order can be paid")
	}
	o.Status = StatusPaid
	o.UpdatedAt = time.Now()
	return nil
}

func (o *Order) Ship() error {
	if o.Status != StatusPaid {
		return errors.New("only paid order can be shipped")
	}
	o.Status = StatusShipped
	o.UpdatedAt = time.Now()
	return nil
}

func (o *Order) Deliver() error {
	if o.Status != StatusShipped {
		return errors.New("only shipped order can be delivered")
	}
	o.Status = StatusDelivered
	o.UpdatedAt = time.Now()
	return nil
}

func (o *Order) Cancel() error {
	if o.Status != StatusPending {
		return errors.New("only pending order can be cancelled")
	}
	o.Status = StatusCancelled
	o.UpdatedAt = time.Now()
	return nil
}

func (o *Order) Refund() error {
	if o.Status != StatusPaid && o.Status != StatusShipped {
		return errors.New("only paid or shipped order can be refunded")
	}
	o.Status = StatusRefunded
	o.UpdatedAt = time.Now()
	return nil
}

File: practice-projects/django-ddd/src/order/domain/order.py
"""Order 聚合根（纯 Python）。

保留一套等价于 gin-ddd / nestjs-ddd 的状态机：
  PENDING → PAID → SHIPPED → DELIVERED
  PENDING → CANCELLED
  PAID|SHIPPED → REFUNDED
"""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from decimal import Decimal
from enum import Enum

from shared.infrastructure.exceptions import DomainError, ValidationError


class OrderStatus(str, Enum):
    PENDING = "PENDING"
    PAID = "PAID"
    SHIPPED = "SHIPPED"
    DELIVERED = "DELIVERED"
    CANCELLED = "CANCELLED"
    REFUNDED = "REFUNDED"


# 允许的状态迁移（显式声明，便于状态机可视化与测试）
_ALLOWED_TRANSITIONS: dict[OrderStatus, set[OrderStatus]] = {
    OrderStatus.PENDING: {OrderStatus.PAID, OrderStatus.CANCELLED},
    OrderStatus.PAID: {OrderStatus.SHIPPED, OrderStatus.REFUNDED},
    OrderStatus.SHIPPED: {OrderStatus.DELIVERED, OrderStatus.REFUNDED},
    OrderStatus.DELIVERED: set(),
    OrderStatus.CANCELLED: set(),
    OrderStatus.REFUNDED: set(),
}


@dataclass
class Order:
    order_no: str
    user_id: int
    total_amount: Decimal
    status: OrderStatus = OrderStatus.PENDING
    id: int | None = None
    created_at: datetime = field(default_factory=datetime.utcnow)
    updated_at: datetime = field(default_factory=datetime.utcnow)

    # ---- 工厂 -----------------------------------------------------------
    @classmethod
    def create(cls, order_no: str, user_id: int, total_amount: Decimal | float | str) -> "Order":
        if not order_no:
            raise ValidationError("订单号不能为空")
        if user_id <= 0:
            raise ValidationError("用户 ID 无效")
        amount = Decimal(str(total_amount))
        if amount <= 0:
            raise ValidationError("订单金额必须大于 0")
        now = datetime.utcnow()
        return cls(
            order_no=order_no,
            user_id=user_id,
            total_amount=amount,
            status=OrderStatus.PENDING,
            created_at=now,
            updated_at=now,
        )

    # ---- 行为（全部走 _transition 校验） --------------------------------
    def pay(self) -> None:
        self._transition(OrderStatus.PAID, "只有待支付订单可以支付")

    def ship(self) -> None:
        self._transition(OrderStatus.SHIPPED, "只有已支付订单可以发货")

    def deliver(self) -> None:
        self._transition(OrderStatus.DELIVERED, "只有已发货订单可以确认送达")

    def cancel(self) -> None:
        self._transition(OrderStatus.CANCELLED, "只有待支付订单可以取消")

    def refund(self) -> None:
        self._transition(OrderStatus.REFUNDED, "只有已支付或已发货的订单可以退款")

    # ---- 查询 -----------------------------------------------------------
    def can_cancel(self) -> bool:
        return self.status == OrderStatus.PENDING

    # ---- 内部 -----------------------------------------------------------
    def _transition(self, target: OrderStatus, err_msg: str) -> None:
        if target not in _ALLOWED_TRANSITIONS.get(self.status, set()):
            raise DomainError(err_msg)
        self.status = target
        self.updated_at = datetime.utcnow()
